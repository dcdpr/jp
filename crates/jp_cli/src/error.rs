use std::io;

use crate::cmd;

pub(crate) type Result<T> = std::result::Result<T, Error>;

/// CLI Error types
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Command error: {0}")]
    Command(#[from] cmd::Error),

    #[error("Config error: {0}")]
    Config(#[from] jp_config::Error),

    #[error("Workspace error: {0}")]
    Workspace(#[from] jp_workspace::Error),

    #[error("Conversation error: {0}")]
    Conversation(#[from] jp_conversation::Error),

    #[error("MCP error: {0}")]
    Mcp(#[from] jp_mcp::Error),

    #[error("LLM error: {0}")]
    Llm(#[from] jp_llm::Error),

    #[error("{0} not found: {1}")]
    NotFound(&'static str, String),

    #[error("IO error: {0}")]
    Io(#[from] io::Error),

    #[error("URL error: {0}")]
    Url(#[from] url::ParseError),

    #[error("Bat error: {0}")]
    Bat(#[from] bat::error::Error),

    #[error("Attachment error: {0}")]
    Attachment(String),

    #[error("Editor error: {0}")]
    Editor(String),

    #[error("Task error: {0}")]
    Task(Box<dyn std::error::Error + Send + Sync>),
}
