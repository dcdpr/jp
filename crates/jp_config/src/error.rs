//! Configuration error.

/// Configuration error.
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// Configuration error.
    #[error(transparent)]
    Schematic(#[from] schematic::ConfigError),

    /// A glob iteration error.
    #[error(transparent)]
    Glob(#[from] glob::GlobError),

    /// A glob pattern parsing error.
    #[error(transparent)]
    Pattern(#[from] glob::PatternError),
}
