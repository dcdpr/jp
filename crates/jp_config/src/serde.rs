#![allow(clippy::ref_option)]

use confique::Partial;
use serde::Deserialize as _;

pub(crate) fn is_nested_default_or_empty<T: Partial + PartialEq>(v: &T) -> bool {
    is_nested_default(v) || is_nested_empty(v)
}

pub(crate) fn is_nested_default<T: Partial + PartialEq>(v: &T) -> bool {
    v == &T::default_values()
}

pub(crate) fn is_nested_empty<T: Partial + PartialEq>(v: &T) -> bool {
    v == &T::empty()
}

pub(crate) fn is_none_or_default<T: Default + PartialEq>(v: &Option<T>) -> bool {
    v.as_ref().is_none_or(|v| v == &T::default())
}

pub(crate) fn is_default<T: Default + PartialEq>(v: &T) -> bool {
    v == &T::default()
}

pub(crate) fn de_from_str<'de, D, T, E>(deserializer: D) -> std::result::Result<T, D::Error>
where
    D: serde::Deserializer<'de>,
    T: std::str::FromStr<Err = E>,
    E: std::error::Error,
{
    String::deserialize(deserializer)
        .and_then(|v| T::from_str(&v).map_err(serde::de::Error::custom))
}

pub(crate) fn de_from_str_opt<'de, D, T, E>(
    deserializer: D,
) -> std::result::Result<Option<T>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: std::str::FromStr<Err = E>,
    E: std::error::Error,
{
    Option::<String>::deserialize(deserializer)?
        .map(|v| T::from_str(&v))
        .transpose()
        .map_err(serde::de::Error::custom)
}
