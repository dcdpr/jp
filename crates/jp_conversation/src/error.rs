use crate::ConversationId;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Failed to serialize XML: {0}")]
    XmlSerialization(#[from] quick_xml::SeError),

    #[error("Invalid ID format: {0}")]
    InvalidIdFormat(String),

    #[error("Invalid ID: {0}")]
    Id(#[from] jp_id::Error),

    #[error("Invalid thread: {0}")]
    Thread(String),

    #[error("unknown conversation ID")]
    UnknownId(ConversationId),
}

#[cfg(test)]
impl PartialEq for Error {
    fn eq(&self, other: &Self) -> bool {
        if std::mem::discriminant(self) != std::mem::discriminant(other) {
            return false;
        }

        // Good enough for testing purposes
        format!("{self:?}") == format!("{other:?}")
    }
}
