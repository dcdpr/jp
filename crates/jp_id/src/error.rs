#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Missing prefix: {0}")]
    MissingPrefix(String),

    #[error("Invalid prefix, must be {0}: {1}")]
    InvalidPrefix(&'static str, String),

    #[error("Missing variant and target id")]
    MissingVariantAndTargetId,

    #[error("Missing variant")]
    MissingVariant,

    #[error("Invalid variant, must be [a-z]: {0}")]
    InvalidVariant(char),

    #[error("Unexpected variant, must be {0}: {1}")]
    UnexpectedVariant(char, char),

    #[error("Missing target ID")]
    MissingTargetId,

    #[error("Invalid timestamp format: {0}")]
    InvalidTimestamp(String),

    #[error("Missing global ID")]
    MissingGlobalId,

    #[error("Invalid global ID, must be [a-z]: {0}")]
    InvalidGlobalId(String),
}
