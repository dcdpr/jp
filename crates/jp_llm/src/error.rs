pub(crate) type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("OpenRouter error: {0}")]
    OpenRouter(#[from] jp_openrouter::Error),

    #[error("Conversation error: {0}")]
    Conversation(#[from] jp_conversation::Error),

    #[error("Config error: {0}")]
    Config(#[from] jp_config::Error),

    #[error("Missing environment variable: {0}")]
    MissingEnv(String),

    #[error("Invalid URL: {0}")]
    Url(#[from] url::ParseError),

    #[error("Invalid response received: {0}")]
    InvalidResponse(String),

    #[error("OpenAI client error: {0}")]
    OpenaiClient(#[from] openai_responses::CreateError),

    #[error("OpenAI event error: {0}")]
    OpenaiEvent(Box<reqwest_eventsource::Error>),

    #[error("OpenAI response error: {0:?}")]
    OpenaiResponse(openai_responses::types::response::Error),

    #[error("OpenAI status code error: {:?} - {}", .status_code, .response)]
    OpenaiStatusCode {
        status_code: reqwest::StatusCode,
        response: String,
    },

    #[error("Gemini error: {0}")]
    Gemini(gemini_client_rs::GeminiError),

    #[error("Ollama error: {0}")]
    Ollama(#[from] ollama_rs::error::OllamaError),

    #[error("Missing structured data in response")]
    MissingStructuredData,

    #[error("Unknown model: {0}")]
    UnknownModel(String),

    #[error("Invalid JSON: {0}")]
    Json(#[from] serde_json::Error),

    #[error("Request error: {0}")]
    Request(#[from] reqwest::Error),

    #[error("Anthropic error: {0}")]
    Anthropic(#[from] async_anthropic::errors::AnthropicError),

    #[error("Anthropic request builder error: {0}")]
    AnthropicRequestBuilder(#[from] async_anthropic::types::CreateMessagesRequestBuilderError),

    #[error("request rate limited (retry after {} seconds)", retry_after.unwrap_or_default().as_secs())]
    RateLimit {
        retry_after: Option<std::time::Duration>,
    },

    #[error("Failed to serialize XML")]
    XmlSerialization(#[from] quick_xml::SeError),

    #[error(transparent)]
    ModelIdConfig(#[from] jp_config::model::id::ModelIdConfigError),

    #[error(transparent)]
    ModelId(#[from] jp_config::model::id::ModelIdError),
}

impl From<gemini_client_rs::GeminiError> for Error {
    fn from(error: gemini_client_rs::GeminiError) -> Self {
        use gemini_client_rs::GeminiError;

        match &error {
            GeminiError::Api(api) if api.get("status").is_some_and(|v| v.as_u64() == Some(404)) => {
                if let Some(model) = api.pointer("/message/error/message").and_then(|v| {
                    v.as_str().and_then(|s| {
                        s.contains("Call ListModels").then(|| {
                            s.split('/')
                                .nth(1)
                                .and_then(|v| v.split(' ').next())
                                .unwrap_or("unknown")
                        })
                    })
                }) {
                    return Self::UnknownModel(model.to_owned());
                }
                Self::Gemini(error)
            }
            _ => Self::Gemini(error),
        }
    }
}

impl From<openai_responses::types::response::Error> for Error {
    fn from(error: openai_responses::types::response::Error) -> Self {
        Self::OpenaiResponse(error)
    }
}

#[cfg(test)]
impl PartialEq for Error {
    fn eq(&self, other: &Self) -> bool {
        if std::mem::discriminant(self) != std::mem::discriminant(other) {
            return false;
        }

        // Good enough for testing purposes
        format!("{self:?}") == format!("{other:?}")
    }
}

#[derive(Debug, thiserror::Error)]
pub enum ToolError {
    #[error("Tool not found")]
    NotFound { name: String },

    #[error("Tools not found")]
    NotFoundN { names: Vec<String> },

    #[error("Disabled in configuration")]
    Disabled,

    #[error("Command is only supported for local tools")]
    UnexpectedCommand,

    #[error("Command missing for local tool")]
    MissingCommand,

    #[error("Failed to fetch tool from MCP client")]
    McpGetToolError(#[source] jp_mcp::Error),

    #[error("Failed to run tool from MCP client")]
    McpRunToolError(#[source] jp_mcp::Error),

    #[error("Failed to serialize tool arguments")]
    SerializeArgumentsError {
        arguments: serde_json::Value,
        #[source]
        error: serde_json::Error,
    },

    #[error("Failed to open editor to edit tool call")]
    OpenEditorError {
        arguments: serde_json::Value,
        #[source]
        error: open_editor::errors::OpenEditorError,
    },

    #[error("Failed to edit tool call")]
    EditArgumentsError {
        arguments: serde_json::Value,
        #[source]
        error: serde_json::Error,
    },

    #[error("Template error")]
    TemplateError {
        data: String,
        #[source]
        error: minijinja::Error,
    },

    #[error("Invalid `type` property for {key}, got {value:?}, expected one of {need:?}")]
    InvalidType {
        key: String,
        value: serde_json::Value,
        need: Vec<&'static str>,
    },

    #[error("Needs input: {question:?}")]
    NeedsInput { question: jp_tool::Question },

    #[error("Serialization error")]
    Serde(#[from] serde_json::Error),

    #[error("Invalid arguments (missing: {missing:?}, unknown: {unknown:?})")]
    Arguments {
        /// Required arguments that were missing.
        missing: Vec<String>,

        /// Unknown arguments that were provided.
        unknown: Vec<String>,
    },
}

#[cfg(test)]
impl PartialEq for ToolError {
    fn eq(&self, other: &Self) -> bool {
        if std::mem::discriminant(self) != std::mem::discriminant(other) {
            return false;
        }

        // Good enough for testing purposes
        format!("{self:?}") == format!("{other:?}")
    }
}
