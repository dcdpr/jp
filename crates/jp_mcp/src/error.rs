use crate::config::McpServerId;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Service error: {0}")]
    Service(#[from] rmcp::ServiceError),

    #[error("MCP error: {0}")]
    Mcp(#[from] rmcp::Error),

    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    #[error("Timeout error: {0}")]
    Timeout(#[from] tokio::time::error::Elapsed),

    #[error("Unknown tool: {0}")]
    UnknownTool(String),

    #[error("Unknown MCP server: {0}")]
    UnknownServer(McpServerId),

    #[error("Invalid tool choice: {0}, must be one of [auto, none, required, fn:<name>]")]
    UnknownToolChoice(String),
}

#[cfg(test)]
impl PartialEq for Error {
    fn eq(&self, other: &Self) -> bool {
        if std::mem::discriminant(self) != std::mem::discriminant(other) {
            return false;
        }

        // Good enough for testing purposes
        format!("{self:?}") == format!("{other:?}")
    }
}
