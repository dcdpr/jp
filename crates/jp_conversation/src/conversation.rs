//! Defines the Conversation structure.

use std::{fmt, str::FromStr};

use jp_config::{Partial, PartialConfig};
use jp_id::{
    parts::{GlobalId, TargetId, Variant},
    Id, NANOSECONDS_PER_DECISECOND,
};
use serde::{ser::SerializeStruct as _, Deserialize, Serialize};
use time::UtcDateTime;

use crate::error::{Error, Result};

/// A sequence of messages between the user and LLM.
#[derive(Debug, Clone, PartialEq, Deserialize)]
pub struct Conversation {
    /// The optional title of the conversation.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,

    /// The last time the conversation was activated.
    pub last_activated_at: UtcDateTime,

    /// Whether the conversation is stored in the user or workspace storage.
    #[serde(skip)]
    pub user: bool,

    /// The partial configuration persisted for this conversation.
    #[serde(default = "PartialConfig::empty")]
    config: PartialConfig,
}

impl Serialize for Conversation {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::Error as _;

        let mut n = 3;
        if self.title.is_some() {
            n += 1;
        }

        let mut state = serializer.serialize_struct("Conversation", n)?;

        if let Some(title) = &self.title {
            state.serialize_field("title", title)?;
        }

        state.serialize_field("last_activated_at", &self.last_activated_at)?;
        state.serialize_field("local", &self.user)?;

        state.serialize_field(
            "config",
            &trim_config(self.config.clone()).map_err(S::Error::custom)?,
        )?;

        state.end()
    }
}

impl Default for Conversation {
    fn default() -> Self {
        Self {
            last_activated_at: UtcDateTime::now(),
            title: None,
            config: PartialConfig::default_values(),
            user: false,
        }
    }
}

impl Conversation {
    #[must_use]
    pub fn new(title: impl Into<String>) -> Self {
        Self {
            title: Some(title.into()),
            ..Default::default()
        }
    }

    #[must_use]
    pub fn with_local(mut self, local: bool) -> Self {
        self.user = local;
        self
    }

    #[must_use]
    pub fn config(&self) -> &PartialConfig {
        &self.config
    }

    #[must_use]
    pub fn config_mut(&mut self) -> &mut PartialConfig {
        &mut self.config
    }

    pub fn set_config(&mut self, config: PartialConfig) {
        self.config = config;
    }
}

/// ID wrapper for Conversation
#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
pub struct ConversationId(#[serde(with = "jp_id::serde")] UtcDateTime);

impl fmt::Debug for ConversationId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_tuple("ConversationId")
            .field(&self.to_string())
            .finish()
    }
}

impl ConversationId {
    #[must_use]
    pub fn timestamp(&self) -> UtcDateTime {
        self.0
    }

    #[must_use]
    pub fn as_deciseconds(&self) -> i128 {
        self.timestamp().unix_timestamp_nanos() / i128::from(NANOSECONDS_PER_DECISECOND)
    }

    pub fn try_from_deciseconds(deciseconds: i128) -> Result<Self> {
        let timestamp = UtcDateTime::from_unix_timestamp_nanos(
            deciseconds * i128::from(NANOSECONDS_PER_DECISECOND),
        )
        .map_err(|e| jp_id::Error::InvalidTimestamp(e.to_string()))?;

        Ok(Self(timestamp))
    }

    pub fn try_from_deciseconds_str(deciseconds: impl AsRef<str>) -> Result<Self> {
        let deciseconds = deciseconds.as_ref().parse::<i128>().map_err(|_| {
            Error::InvalidIdFormat(format!("Invalid deciseconds: {}", deciseconds.as_ref()))
        })?;

        Self::try_from_deciseconds(deciseconds)
    }

    pub fn from_dirname(dirname: impl AsRef<str>) -> Result<Self> {
        dirname
            .as_ref()
            .split('-')
            .next()
            .ok_or(jp_id::Error::MissingTargetId.into())
            .and_then(Self::try_from_deciseconds_str)
    }

    pub fn to_dirname(&self, title: Option<&str>) -> Result<String> {
        let len = title.map(str::len).unwrap_or_default();
        let title = title
            .unwrap_or_default()
            .trim()
            .chars()
            .take(60)
            .map(|c| if c.is_alphanumeric() { c } else { '-' })
            .collect::<String>()
            .trim_end_matches('-')
            .to_lowercase();

        let ts = self.as_deciseconds().to_string();
        if title.is_empty() {
            return Ok(ts);
        }

        let mut title = format!("{ts}-{title}");
        if len > 60
            && let Some(i) = title.rfind('-')
        {
            title.truncate(i);
        }

        Ok(title)
    }
}

impl TryFrom<UtcDateTime> for ConversationId {
    type Error = Error;

    fn try_from(timestamp: UtcDateTime) -> Result<Self> {
        let nanos = timestamp.nanosecond();
        let truncated_nanos = (nanos / NANOSECONDS_PER_DECISECOND) * NANOSECONDS_PER_DECISECOND;

        timestamp
            .replace_nanosecond(truncated_nanos)
            .map_err(|e| jp_id::Error::InvalidTimestamp(e.to_string()).into())
            .map(Self)
    }
}

impl Id for ConversationId {
    fn variant() -> Variant {
        'c'.into()
    }

    fn target_id(&self) -> TargetId {
        self.as_deciseconds().to_string().into()
    }

    fn global_id(&self) -> GlobalId {
        jp_id::global::get().into()
    }
}

impl fmt::Display for ConversationId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.format_id(f)
    }
}

impl FromStr for ConversationId {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        jp_id::parse::<Self>(s)
            .map(|p| p.target_id)
            .map_err(Into::into)
            .and_then(Self::try_from_deciseconds_str)
    }
}

impl Default for ConversationId {
    fn default() -> Self {
        Self::try_from(UtcDateTime::now()).expect("valid timestamp")
    }
}

/// Holds metadata about all conversations, like the current active
/// conversation.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConversationsMetadata {
    /// The ID of the currently active conversation.
    ///
    /// If no active conversation exists, one is created.
    pub active_conversation_id: ConversationId,
}

impl ConversationsMetadata {
    #[must_use]
    pub fn new(active_conversation_id: ConversationId) -> Self {
        Self {
            active_conversation_id,
        }
    }
}

impl Default for ConversationsMetadata {
    fn default() -> Self {
        Self::new(ConversationId::default())
    }
}

fn trim_config(
    cfg: jp_config::PartialConfig,
) -> std::result::Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
    Ok(match serde_json::to_value(cfg)? {
        serde_json::Value::Object(v) => v
            .into_iter()
            .filter_map(|(k, v)| compact_value(v).map(|v| (k, v)))
            .collect::<serde_json::Map<String, serde_json::Value>>()
            .into(),
        v => v,
    })
}

fn compact_value(v: serde_json::Value) -> Option<serde_json::Value> {
    match v {
        serde_json::Value::Object(v) => {
            let map = v
                .into_iter()
                .filter_map(|(k, v)| compact_value(v).map(|v| (k, v)))
                .collect::<serde_json::Map<_, _>>();

            (!map.is_empty()).then(|| map.into())
        }
        serde_json::Value::Array(v) => {
            let vec = v.into_iter().filter_map(compact_value).collect::<Vec<_>>();

            (!vec.is_empty()).then(|| vec.into())
        }
        serde_json::Value::Null => None,
        _ => Some(v),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_conversation_serialization() {
        let mut config = PartialConfig::default_values();
        config.assistant.provider.openai.base_url = None;
        config.assistant.provider.openrouter.app_name = None;
        config.style.code = <_>::empty();

        let conv = Conversation {
            title: None,
            last_activated_at: UtcDateTime::new(
                time::Date::from_calendar_date(2023, time::Month::January, 1).unwrap(),
                time::Time::MIDNIGHT,
            ),
            user: true,
            config,
        };

        insta::assert_json_snapshot!(conv);
    }
}
