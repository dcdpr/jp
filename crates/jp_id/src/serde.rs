use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
use time::UtcDateTime;

use crate::NANOSECONDS_PER_DECISECOND;

/// Serialize an `UtcDateTime` as its Unix timestamp with deciseconds.
pub fn serialize<S: Serializer>(
    datetime: &UtcDateTime,
    serializer: S,
) -> std::result::Result<S::Ok, S::Error> {
    let timestamp = datetime.unix_timestamp_nanos() / i128::from(NANOSECONDS_PER_DECISECOND);
    timestamp.serialize(serializer)
}

/// Deserialize an `UtcDateTime` from its Unix timestamp with deciseconds.
pub fn deserialize<'a, D: Deserializer<'a>>(
    deserializer: D,
) -> std::result::Result<UtcDateTime, D::Error> {
    let num = i128::deserialize(deserializer)?;
    UtcDateTime::from_unix_timestamp_nanos(num * i128::from(NANOSECONDS_PER_DECISECOND))
        .map_err(de::Error::custom)
}
