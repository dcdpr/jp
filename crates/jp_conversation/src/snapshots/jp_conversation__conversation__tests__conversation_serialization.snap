---
source: crates/jp_conversation/src/conversation.rs
expression: conv
---
{
  "last_activated_at": "2023-01-01 00:00:00.0",
  "local": true
}
